tests/unit/peapods/runtimes/gateway/test_prefetch.py

```diff
@@ -0,0 +1,49 @@
+import asyncio
+from asyncio import Future
+
+import pytest
+
+from jina.helper import ArgNamespace, random_identity
+from jina.parsers import set_gateway_parser
+from jina.peapods.runtimes.gateway.prefetch import PrefetchCaller
+from jina.proto import jina_pb2
+
+
+def _generate_request():
+    req = jina_pb2.RequestProto()
+    req.request_id = random_identity()
+    req.data.docs.add()
+    return req
+
+
+class ZmqletMock:
+    def __init__(self):
+        self.sent_future = Future()
+        self.received_event = asyncio.Event()
+
+    async def recv_message(self, **kwargs):
+        msg = await self.sent_future
+        self.sent_future = Future()
+        self.received_event.set()
+        return msg
+
+    async def send_message(self, message):
+        self.sent_future.set_result(_generate_request())
+        await self.received_event.wait()
+        self.sent_future.set_result(message.response)
+
+
+@pytest.mark.asyncio
+async def test_concurrent_requests():
+    args = ArgNamespace.kwargs2namespace({}, set_gateway_parser())
+    mock_zmqlet = ZmqletMock()
+    servicer = PrefetchCaller(args, mock_zmqlet)
+
+    request = _generate_request()
+
+    response = servicer.send(iter([request]))
+
+    async for r in response:
+        assert r == request
```

Comment:
Shouldn't we send a big bunch of requests for testing the beforehand failing behavior?

Reply:
I am simulating this behaviour with the mock here: https://github.com/jina-ai/jina/pull/2843/files#diff-4836b2c1886183887896c31a1d41c18e6e6a3f8f371e10ea5bd04d7563cf2c8bR31

This basically will respond to a request first with some random unrelated response and then with the correct one. This fails with current master.

Sending actually a lot of other messages is in the test a bit complicated and I would rather avoid in this unit test (I would need to create threads etc). This scenario is also rather well tested with the real end to end test of Felix.